probe-rs/src/flashing/builder.rs

```diff
@@ -0,0 +1,697 @@
+                            }
+                            continue;
+                        } else {
+                            let space_left_in_page = page.size - page.data.len() as u32;
+                            let space_left_in_data = op.data.len() - pos;
+                            let amount =
+                                usize::min(space_left_in_page as usize, space_left_in_data);
+
+                            page.data.extend(&op.data[pos..pos + amount]);
+                            log::trace!("Added {} bytes to current page", amount);
+                            pos += amount;
+                        }
+                    } else {
+                        // If no page is on the sector yet.
+                        let page_info = flash_algorithm.page_info(flash_address);
+                        if let Some(page_info) = page_info {
+                            let new_page = FlashPage::new(&page_info);
+                            pages.push(new_page.clone());
+                            log::trace!(
+                                "Added Page (0x{:08x}..0x{:08x})",
+                                page_info.base_address,
+                                page_info.base_address + page_info.size
+                            );
+                        } else {
+                            return Err(FlashError::InvalidFlashAddress(flash_address));
+                        }
+                        continue;
+                    }
+                } else {
+                    // If no sector exists, create a new one.
+                    log::trace!("Trying to create a new sector");
+                    let sector_info = flash_algorithm.sector_info(flash_address);
+
+                    if let Some(sector_info) = sector_info {
+                        let new_sector = FlashSector::new(&sector_info);
+                        sectors.push(new_sector);
+                        log::debug!(
+                            "Added Sector (0x{:08x}..0x{:08x})",
+                            sector_info.base_address,
+                            sector_info.base_address + sector_info.size
+                        );
+                    } else {
+                        return Err(FlashError::InvalidFlashAddress(flash_address));
+                    }
+                    continue;
+                }
+            }
+        }
+
+        // Fill the page gap if there is one.
+        if let Some(page) = pages.last_mut() {
+            fill_page(page)?;
+        }
+
+        log::debug!("Sectors are:");
+        for sector in &sectors {
+            log::debug!("{:#?}", sector);
+        }
+
+        log::debug!("Pages are:");
+        for page in &pages {
+            log::debug!("{:#?}", page);
+        }
+
+        Ok(FlashLayout {
+            sectors,
+            pages,
+            fills: vec![],
+        })
+    }
+
+    pub(super) fn add_sector<'b>(
+        &self,
```

Comment:
It seems that self is never used in the function. Could we put the sectors into the builder itself, instead of a parameter?

Reply:
I would rather keep the sectors locally as they are only created in one method (`build_sectors_and_pages`) and returned from that.
We could remove self as a param tho.